doc/compiler/GPU/README.md

```diff
@@ -0,0 +1,228 @@
+
+![An overview of the GPU code generation.](./figs/gpucodegen.png "An overview of the GPU code generation." )
+
+When the JIT compiler identifies the parallel stream API in the IR, it first applies a set of GPU-aware optimizations, then generates [NVVM IR](https://docs.nvidia.com/cuda/nvvm-ir-spec/index.html), and eventually generates a GPU kernel binary through [the ```libNVVM``` library](https://docs.nvidia.com/cuda/libnvvm-api/index.html) as well as the host CPU code (```GPU Runtime``` in the figure) that is responsible for device memory allocations, data transfers between the host and the device, and kernel invocations.
+
+For more details, refer to our paper on compiling and optimizing Java 8 programs for GPU execution: [link](https://ieeexplore.ieee.org/document/7429325).
+
+## How to build OpenJ9 with the GPU code generation enabled
+To build OpenJ9 with the GPU code generation enabled, it requires a few more steps than [the original OpenJ9 build instruction](https://www.eclipse.org/openj9/oj9_build.html). Essentially, you need to 1) modify the Dockerfile to build a GPU-aware docker image, and 2) give a few variables when building OpenJDK with OpenJ9 (when doing ```make all```). The followings describe the steps to building OpenJ9 with the GPU code generation enabled:
+
+### Prepare your system
+
+#### Enable a GPU-aware docker image
+First of all, when preparing your system, make sure that you have installed [the NVIDIA Container Runtime for Docker](https://github.com/NVIDIA/nvidia-docker) as well as [the NVIDIA driver](https://github.com/NVIDIA/nvidia-docker/wiki/Frequently-Asked-Questions#how-do-i-install-the-nvidia-driver) and [a supported version of Docker](https://github.com/NVIDIA/nvidia-docker/wiki/Frequently-Asked-Questions#which-docker-packages-are-supported).
+
+Before proceeding to the next step, it is a good idea to make sure the installations are successfull. Let's make sure that your GPUs are recognized and the CUDA compiler ```nvcc``` is available within a docker container. In the following example, we create a docker container based on the ```nvidia/cuda:9.2-devel-ubuntu16.04``` image, and perform two commands: 1) the ```nvidia-smi``` command, which gives you detailed information on NVIDIA GPU devices, and 2) the ```nvcc``` command with the ```--version``` option. In this case, ```nvidia-smi``` shows the NVIDIA driver version 410.48 and the TITAN Xp GPU are found. Also, the version of the ```nvcc``` compiler is 9.2.
+
+**Note:** you can pick any CUDA version that is supported in your GPU, but always make sure you pick a ```devel``` image because the GPU code generation feature requires the CUDA compiler toolchain. A full list of the available CUDA docker images can be found at [the Docker Hub](https://hub.docker.com/r/nvidia/cuda/). For ```ppc64le``` users, see [here](https://hub.docker.com/r/nvidia/cuda-ppc64le/).
```

Comment:
For the scenarios discussed here, I don't think you will actually be using `nvcc`. Only the CUDA toolkit header files are required for building the JVM. At runtime, only the toolkit and driver shared libraries are needed; JIT generates IR code for `nvvm` which transforms it into GPU code. The code underlying the acceleration of `Arrays.sort()` is generated as PTX code which can be loaded directly by the driver.

Reply:
Right, here I used ```nvcc``` only for checking purposes. I'll update the text just to check required things such as ```nvvm.h``` and so on.